src/Protocols/Wishbone/Standard/Hedgehog.hs

```diff
@@ -25,63 +27,184 @@ import Protocols hiding (stallC, circuit)
-  | busCycle m2s && P.not (strobe m2s) = Right WSSInCycleNoStrobe
-  | not (busCycle m2s) = Right WSSQuiet
-  | acknowledge s2m || err s2m || retry s2m = Right WSSQuiet
-  | otherwise = Right WSSWaitForSlave
+    Left LVETerminationSignalOutsideOfCycle
+nextStateLenient LVSQuiet m2s _
+  | busCycle m2s && P.not (strobe m2s) = Right LVSInCycleNoStrobe
+  | busCycle m2s && strobe m2s = Right LVSWaitForSlave
+  | otherwise = Right LVSQuiet
+nextStateLenient LVSInCycleNoStrobe m2s _
+  | not (busCycle m2s) = Right LVSQuiet
+  | busCycle m2s && strobe m2s = Right LVSWaitForSlave
+  | otherwise = Right LVSInCycleNoStrobe
+nextStateLenient LVSWaitForSlave m2s s2m
+  | busCycle m2s && P.not (strobe m2s) = Right LVSInCycleNoStrobe
+  | not (busCycle m2s) = Right LVSQuiet
+  | acknowledge s2m || err s2m || retry s2m = Right LVSQuiet
+  | otherwise = Right LVSWaitForSlave
 
--- | Validate the input/output streams of a standard wishbone interface
+-- | Lenient validation of a wishbone-component based on the specification
 --
 -- Returns 'Right ()' when the interactions comply with the spec.
 -- Returns 'Left (cycle, err)' when a spec violation was found.
-validateStandard ::
+validateLenient ::
   [WishboneM2S addressWidth (BitSize a `DivRU` 8) a] ->
   [WishboneS2M a] ->
-  Either (Int, WishboneStandardError) ()
-validateStandard m2s s2m = go 0 (P.zip m2s s2m) WSSQuiet
+  Either (Int, LenientValidationError) ()
+validateLenient m2s s2m = go 0 (P.zip m2s s2m) LVSQuiet
   where
     go _ [] _ = Right ()
-    go n ((fwd, bwd) : rest) st0 = case nextStateStandard st0 fwd bwd of
+    go n ((fwd, bwd) : rest) st = case nextStateLenient st fwd bwd of
       Left e -> Left (n, e)
-      Right st1 -> go (n + 1) rest st1
+      Right st' -> go (n + 1) rest st'
+
+--
+-- Validation for "common sense" compliance
+--
+-- The Wishbone spec itself makes very little assumptions about how interactions
+-- should look like outside of basic hand-shaking.
+-- This "common sense" compliance checking additionally checks for:
+--  - A read request is acked with defined data
+--    - response data respects the 'busSelect' signal
+--  - A write request is answered with only an ACK/ERR/RETRY and no defined data
+--    - components might want to reply with data
+--
+
+type CommonSenseValidationError = String
+
+data CommonSenseValidationState
+  = CSVSQuiet
+  | CSVSInCycleNoStrobe
+  | CSVSReadCycle
+  | CSVSWriteCycle
+  deriving (Eq, Show)
+
+nextStateCommonSense ::
+  forall a addressWidth.
+  (BitPack a) =>
+  CommonSenseValidationState ->
+  WishboneM2S addressWidth (BitSize a `DivRU` 8) a ->
+  WishboneS2M a ->
+  Either CommonSenseValidationError (Bool, CommonSenseValidationState)
+--                                    ^ go to next cycle
+nextStateCommonSense _ _ s2m
+  | P.length (filter ($ s2m) [acknowledge, err, retry]) > 1 =
+    Left "More than one termination signal asserted"
+nextStateCommonSense CSVSQuiet WishboneM2S {..} s2m
```

Comment:
Style nitpick: wouldn't it make more sense to use a case here instead of repeating `nextStateCommonSense CSVSQuiet WishboneM2S {..} s2m` a bunch of times?

Reply:
Remove the space between `WishboneM2S` and `{..}`; this makes it clearer that these really belong together instead of being multiple arguments.

Reply:
Also if you feel aligning stuff is justified, feel free. I didn't mean to crush alignment altogether with my earlier comments :-).

Reply:
I can try with a `case` instead, the guards will just create longer lines then, or I'll need `MultiWayIf`. I'll give it a try!

Reply:
Should I do the same for the `nextStateLenient` function or is that small enough that it's not as significant?

Reply:
Changed that function too for local consistency :+1: